vignettes/articles/managing-tokens-securely.Rmd

```diff
@@ -0,0 +1,198 @@
+
+### Provide env var to other services
+
+On Travis-CI, define this env var in your repo settings via the browser UI:
+
+https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings
+
+Alternatively, you can use the Travis command line interface to configure the env var or even define an encrypted env var in `.travis.yml`. In any case, remember that private env vars are not available to pull requests, which is another reason to carry on gracefully when token decryption is not possible (see below).
+
+On R-hub, send this env var in your calls to `check()` and friends:
+
+```
+rhub::check(env_vars = Sys.getenv(secret_pw_name("gargle"), names = TRUE))
+```
+
+### Encrypt the secret file
+
+`secret_write()` takes 3 arguments:
+
+  * `package` name. Processed through `secret_pw_name()` in order to retrieve
+    the PASSWORD from an appropriately named env var.
+  * `name` of the encrypted file to write. The location is below `inst/secret`
+    in the source of `package`.
+  * `data`, either a file path to the unencrypted secret file or the data to
+    be encrypted as a raw vector. In the case of a secret file, it should
+    normally be stored outside your package, to decrease the chance of
+    accidental leakage.
+
+Example of a call to `secret_write()`, where `gargle-testing.json` is a JSON
+file downloaded for a service account managed via the [Google API / Cloud Platform console](https://console.cloud.google.com/project).
+
+```{r eval = FALSE}
+secret_write(
+  package = "gargle",
+  name = "gargle-testing.json",
+  data = "~/.R/gargle/gargle-testing.json"
+)
+```
+
+### Test setup
+
+Now you need to rig your tests or their setup around this encrypted token. You need to plan for two scenarios:
+
+  * Decryption is not going to work. Either because the Suggested [sodium](https://cran.r-project.org/web/packages/sodium/index.html) package is not available or (much more likely) because the env var that represents the key is not available.
+    - This will be the case on CRAN, by definition, because there is no way to share an encrypted secret.
+    - This will be the case for external contributors, on their personal machines and when their GitHub pull requests are checked via CI services, such as Travis-CI or AppVeyor. 
+  * Decryption is going to work. This is where you actually get to test package functionality against the target API.
+
+In a wrapper package, you could determine decrypt-ability at the start of the test run. Here's representative code from bigrquery's `tests/testthat/setup-auth.R` file:
+
+```{r eval = FALSE}
+if (secret_can_decrypt("bigrquery")) {
```

Comment:
`gargle:::secret_can_decrypt` and `gargle:::secret_read`

Reply:
In this case, no, because it really does refer to versions of `secret_can_decrypt()` and `secret_read()` that are available in bigrquery. Which is how other wrappers would probably have to do it too because I bet `:::` would cause trouble at CRAN. I've added explanatory words.